core/jobs/batch_jobs/exp_validation_jobs.py

```diff
@@ -0,0 +1,298 @@
+          interaction having answer group(s) with non-empty training data.
+    """
+
+    def get_states_having_invalid_state_classifier_id(self, states):
+        """Returns a list of state names that have not None
+        classifier model id.
+
+        Args:
+            states: list[States]. The list of states.
+
+        Returns:
+            list[str]. Returns a list of state names that have not None
+            classifier model id.
+        """
+        invalid_state_names = []
+        for state_name in states:
+            if states[state_name].classifier_model_id is not None:
+                invalid_state_names.append(state_name)
+        return invalid_state_names
+
+    def filter_exps_having_invalid_state_classifier_model_id(self, states):
+        """Returns True if any state has classifier model id not None.
+
+        Args:
+            states: dict(str, State). A dict of states.
+
+        Returns:
+            bool. Returns True if any state has classifier model id not None.
+        """
+        return any(
+            state.classifier_model_id is not None for state in states.values()
+        )
+
+    def get_states_having_invalid_training_data(self, exploration):
+        """Returns state names having non-empty training data.
+
+        Args:
+            exploration: Exploration. Exploration to be checked.
+
+        Returns:
+            list(str). Returns state names having non-empty training data.
+        """
+        state_names = []
+        for state_name in exploration.states:
+            for answer_group in exploration.states[
+                state_name
+            ].interaction.answer_groups:
+                if len(answer_group.training_data) > 0:
+                    state_names.append(state_name)
+                    # The below break is necessary to avoid duplicated
+                    # state names in the list.
+                    break
+        return state_names
+
+    def filter_curated_explorations(self, model_pair):
+        """Returns True if the exploration is curated.
+
+        Args:
+            model_pair: tuple. The pair of exp and opportunity models.
+
+        Returns:
+            bool. Returns True if opportinity model exists for the given id.
+        """
+        exp_and_opportunity_models = list(model_pair[1])
+        return len(exp_and_opportunity_models) == 2
+
+    def get_exploration_from_model_pair(self, model_pair):
+        """Returns the exploration from the model pair.
+
+        Args:
+            model_pair: tuple. The pair of exp and opportunity models.
+
+        Returns:
+            Exploration. The exploration domain object.
+        """
+        exp_and_opportunity_models = list(model_pair[1])
+        exp_model = exp_and_opportunity_models[0]
+        if not isinstance(exp_model, exp_models.ExplorationModel):
+            exp_model = exp_and_opportunity_models[1]
+        return exp_fetchers.get_exploration_from_model(exp_model)
```

Comment:
This is a bit iffy. Can we somehow make sure that the tuple will always have the same order?

Reply:
I have done this now by COGroupByKey. PTAL.